### PR TITLE
[GrahpBolt] refine hetero_rgcn example

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

#6674 
Refine hetero_rgcn example, aligning it with `examples/sampling/graphbolt/node_classification.py`.

stdout:
```
$ python examples/sampling/graphbolt/rgcn/hetero_rgcn.py --num_gpus 1
Downloading datasets/ogbn-mag.zip from https://data.dgl.ai/dataset/graphbolt/ogbn-mag.zip...
Extracting file to datasets
Start to preprocess the on-disk dataset.
Finish preprocessing the on-disk dataset.
Loaded dataset: node_classification
node_num for rel_graph_embed: {'author': tensor(1134649), 'field_of_study': tensor(59965), 'institution': tensor(8740)}
Number of embedding parameters: 154029312
Number of model parameters: 337460
Start to train...
Evaluating the model on the validation set.
Finish evaluating on validation set.
Epoch: 01, Loss: 2.6244, Valid accuracy: 33.57%, Time 61.9471
Evaluating the model on the validation set.
Finish evaluating on validation set.
Epoch: 02, Loss: 2.0163, Valid accuracy: 35.10%, Time 60.7852
Evaluating the model on the validation set.
Finish evaluating on validation set.
Epoch: 03, Loss: 1.7061, Valid accuracy: 36.70%, Time 60.7745
Testing...
Test accuracy 35.2679
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
update https://github.com/dmlc/dgl/tree/master/examples/sampling/graphbolt/rgcn#accuracies accordingly

**Review by Rhett-Ying — COMMENTED** [review]
1. model.inference()?
2. remove eval on train_set?
3. args such as `--epochs`?

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 1]
```diff
@@ -654,60 +593,41 @@ def main(args):
...
+    # which is passed to the optimizer. The optimizer then updates all
+    # these parameters during the training process.
+    all_params = itertools.chain(model.parameters())
+    optimizer = torch.optim.Adam(all_params, lr=0.01)
+
+    # `expected_max`` is the number of physical cores on your machine.
+    # The `logical` parameter, when set to False, ensures that the count
+    # returned is the number of physical cores instead of logical cores
+    # (which could be higher due to technologies like Hyper-Threading).
+    expected_max = int(psutil.cpu_count(logical=False))
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:610]: remove it for simplicity

**Comment by Skeleton003** [comment]
> 1. model.inference()?
> 2. remove eval on train_set?
> 3. args such as `--epochs`?

https://github.com/dmlc/dgl/blob/master/examples/sampling/graphbolt/rgcn/hetero_rgcn.py#L549-L550
Do we allow number of epochs to be more than 3 in this example?

**Comment by Rhett-Ying** [comment]
@Skeleton003 is this PR ready? integrate the change made in https://github.com/dmlc/dgl/pull/6875 for benchmark?

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 2]
```diff
@@ -298,7 +338,7 @@ def forward(self, g, inputs):
         inputs_dst = {
             k: v[: g.number_of_dst_nodes(k)] for k, v in inputs.items()
         }
-
+        # assert False, (inputs, inputs_dst)
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:341]: remove?

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 3]
```diff
@@ -366,90 +406,45 @@ def reset_parameters(self):
...
-            print(f"Highest Train: {r.mean():.2f} ± {r.std():.2f}")
-            r = best_result[:, 1]
-            print(f"Highest Valid: {r.mean():.2f} ± {r.std():.2f}")
-            r = best_result[:, 2]
-            print(f"  Final Train: {r.mean():.2f} ± {r.std():.2f}")
-            r = best_result[:, 3]
-            print(f"   Final Test: {r.mean():.2f} ± {r.std():.2f}")
+        y = list()
+        for data in tqdm(dataloeader, desc="Model Inference"):
+            blocks = [block.to(device) for block in data.blocks]
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:433]: `datapipe.copy_to()`? for now, copy to GPU is supported.

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 4]
```diff
@@ -716,11 +712,12 @@ def main(args):
...
         type=str,
         default="ogbn-mag",
+        choices=["ogbn-mag", "ogb-lsc-mag240m"],
         help="Dataset name. Possible values: ogbn-mag, ogb-lsc-mag240m",
     )
-    parser.add_argument("--runs", type=int, default=5)
+    parser.add_argument("--num_epochs", type=int, default=3)
     parser.add_argument("--num_workers", type=int, default=0)
-    parser.add_argument("--num_gpus", type=int, default=0)
+    parser.add_argument("--num_gpus", type=int, default=1)
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:720]: run on CPU in default

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 5]
```diff
@@ -580,57 +570,70 @@ def run(
...
-    g, features, train_set, valid_set, test_set, num_classes = load_dataset(
-        args.dataset
-    )
+    (
+        g,
+        features,
+        train_set,
+        valid_set,
+        test_set,
+        all_nodes_set,
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:634]: is this variable used?

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 6]
```diff
@@ -580,57 +570,70 @@ def run(
...
+        name=name,
+        graph=g,
+        features=features,
+        item_set=item_set,
+        node_embed=node_embed,
+        device=device,
+        num_workers=num_workers,
+    )
+    print(pred)
+    label = item_set._itemsets["paper"]._items[1].to(pred.device)
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:613]: access internal data of class object in example is not a good practice.

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
How about this? Let's make below changes(especially `inference()` is excluded) only in this PR. just make it simple:
- remove unnecessary arguments such as `--runs`.
- update the log format for benchmark.
- update device in copy_to() for GPU case.
- update `README` accordingly.

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py (hunk 6 above)
Comment (Skeleton003) [reply examples/sampling/graphbolt/rgcn/hetero_rgcn.py:613]: yes, this PR is not ready yet and these unnecessary lines will be removed.

**Comment by Skeleton003** [comment]
@dgl-bot

File: examples/sampling/graphbolt/rgcn/README.md [hunk 7]
```diff
@@ -24,12 +24,11 @@ Below results are roughly collected from an AWS EC2 **g4dn.metal**, 384GB RAM, 9
...
-Final performance: 
-All runs:
-Highest Train: 64.66 ± 0.74
-Highest Valid: 41.31 ± 0.12
-  Final Train: 64.66 ± 0.74
-   Final Test: 40.07 ± 0.02
+Epoch: 01, Loss: 2.6244, Valid accuracy: 33.57%, Time 61.9471
+Epoch: 02, Loss: 2.0163, Valid accuracy: 35.10%, Time 60.7852
+Epoch: 03, Loss: 1.7061, Valid accuracy: 36.70%, Time 60.7745
+Test accuracy 35.2679
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/README.md:30]: is this the results running with latest GraphBolt code?

Comment (Skeleton003) [reply examples/sampling/graphbolt/rgcn/README.md:30]: > is this the results running with latest GraphBolt code?

yes

File: examples/sampling/graphbolt/rgcn/hetero_rgcn.py [hunk 8]
```diff
@@ -578,59 +563,42 @@ def run(
...
+            f"Time {t1 - t0:.4f}"
         )
-        print("Finish evaluating on test set.")
-
-    return logger
 
 
 def main(args):
-    device = th.device("cuda") if args.num_gpus > 0 else th.device("cpu")
+    if args.dataset == "ogb-lsc-mag240m":
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/rgcn/hetero_rgcn.py:586]: remove it

**Comment by Rhett-Ying** [comment]
@caojy1998 pls check the output format for benchmark

**Review by caojy1998 — APPROVED** [review]
The code seems good to me.

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**